src/mlpack/methods/reinforcement_learning/q_learning.hpp

```diff
@@ -52,7 +53,8 @@ template <
   typename NetworkType,
   typename UpdaterType,
   typename PolicyType,
-  typename ReplayType = RandomReplay<EnvironmentType>
+  typename ReplayType = RandomReplay<EnvironmentType>,
```

Comment:
Maybe I missed something, but I think it's possible to use the existing Interface and implement PrioritizedReplay as another policy, so a user can just pass ``PrioritizedReplay<EnvironmentType>`` instead of ``RandomReplay<EnvironmentType>`` does this make sense?

Reply:
> Maybe I missed something, but I think it's possible to use the existing Interface and implement PrioritizedReplay as another policy, so a user can just pass `PrioritizedReplay<EnvironmentType>` instead of `RandomReplay<EnvironmentType>` does this make sense?

Hi, @zoq, thank for your comments and sorry for the delay because of the Spring Festival. 

Yes, there may many different implementations of this. Changing that Interface is one of my implementations.  Let me clarify my thoughts on this. In my understanding, the policy outputs the agent's action to tell the agent what need to when facing the current state. The replay buffer is the place where the agent to store the experience. The offline RL algorithm can update the parameter by sampling experience from the replay buffer. 

`PrioritizedReplay` is the same as `RandomReplay ` except that the sampling behavior.  Currently, I implement two different `constructor` function, so the user can choose what type of replay buffer to use.  

You mentioned `implement PrioritizedReplay as another policy`, do you mean we put the `replay buffer`'s part into the policy?  I saw that the existing policy in current `policy` folder, all the interface are sampling action. Can you let me know more about your idea?

Reply:
Right, ideally we could put everything into the policy, that way I could use prioritized replay for other classes as well. What about adding another method to the replay class, something like ``Update(...)`` that is called right after ``  updater.Update(learningNetwork.Parameters(), config.StepSize(), gradients);``. For the existing replay classes we do nothing but for the prioritized replay  class we update the priorities. Do you think that could work?

Reply:
> Right, ideally we could put everything into the policy, that way I could use prioritized replay for other classes as well. What about adding another method to the replay class, something like `Update(...)` that is called right after ` updater.Update(learningNetwork.Parameters(), config.StepSize(), gradients);`. For the existing replay classes we do nothing but for the prioritized replay class we update the priorities. Do you think that could work?

Yes, you are right. `... adding another method to the replay class, something like Update(...) that is called right after ...`  it will look like more elegent. 

Idealy, I hope that there is only one `replayMethod` variable instead of one `replayMethod` and one `prioritizedReplayMethod`. If the user pass `random_replay` object, we can sample experience randomly. and sampling with priorities  when pass `prioritized_replay` object. But class `random_replay` and `prioritized_replay` have no inheritence relation, I not sure that they can cast with each other. Do you have any idea about it?

Reply:
> Right, ideally we could put everything into the policy, that way I could use prioritized replay for other classes as well. What about adding another method to the replay class, something like `Update(...)` that is called right after ` updater.Update(learningNetwork.Parameters(), config.StepSize(), gradients);`. For the existing replay classes we do nothing but for the prioritized replay class we update the priorities. Do you think that could work?

Hi, @zoq, I update the code that adding an `update` method for `prioritized_replay` Class to update the priorities. I think the code can satisfy all the functionality of `prioritized_replay`.

BTW, I found that there is a way to make the code more elegant, but it needs C++14 support. 
We can use `boost:: variant` to achieve different behavior when using different replay buffer Class.
 
```
using Obj = boost::variant<RandomReplay, PrioritizedReplay>;
```
If we use RandomReplay, the code can be writen in this way:
```
Obj v = RandomReplay(...);
boost::apply_visitor([](auto const& obj) { obj.Sample(); }, v);
```
or PrioritizedReplay
```
Obj v = PrioritizedReplay(...);
boost::apply_visitor([](auto const& obj) { obj.Sample(); }, v);
```

the reference link: [boost variant simple call to common methods](https://stackoverflow.com/questions/37504740/boost-variant-simple-call-to-common-methods)

Reply:
``boost::variant``  is a really versatile construct, but I don't think we should switch to C++14 yet, but we should keep that in mind. Will take a look at the updates, thanks for looking into the issue.